### ISSUE TITLE
Document non-determinism on the website and/or in the docs

### ISSUE THREAD

**Issue by matt-gardner** [allenai/allennlp#387]
There is significant variance in training runs due to CUDNN.  We can't really fix this, but it should at least be clearly documented wherever it is relevant.

Here's some analysis done by @OyvindTafjord, pasted here from a private email, so we have somewhere public to point to if this comes up again, and some notes for anyone wanting to make a PR with documentation.

----------------

I think I isolated the issue causing our BiDAF on SQuAD training runs to produce different results in spite of fixed random seeds (I have seen 67.7, 67.9, and 68.3 ExactMatch in three different runs, Matt G got 68.7 in another one). It seems to be related to inherent indeterminism in CUDNN code, due to the way things are optimized for speed. As I understand it, asynchronous calculations can cause operations to be done in different orders, so numerical accuracy causes differences which then magnify (to surprising amount?) in big computations. 

The upshot is that there doesn't seem to be a way to avoid this without taking a significant speed penalty (using CUNN instead of CUDNN for GPU computations), but other than causing annoying noise in experiments, there probably isn't much reason to demand such determinism. E.g., although the results are reproducible using CUNN or CPU, they are different between CUNN and CPU (and also between different CUDA versions and CPU OS'es). Rather one should presumably just treat this as actual experimental noise and perhaps try to estimate the uncertainty this gives to the results.

Here are some random mentions of this:

https://github.com/soumith/cudnn.torch/issues/274
https://discuss.pytorch.org/t/choose-backends/271
https://discuss.pytorch.org/t/non-determinisic-results/459/4 
https://github.com/soumith/cudnn.torch/issues/270

To confirm that this is what's going on, I ran a few configurations of BiDAF on a 10% sample of SQuAD (on an AWS p2.xlarge):

1) Default  (CUDNN)
2) Set torch.backends.cudnn.enabled = False so that CUNN is used instead of CUDNN
3) CPU mode  ("cuda_device": -1)

In CUDNN mode, the results are indeed varying between runs, even after one epoch. E.g, after 2 epochs I see validation F1/loss of 15.8/7.20 vs 16.1/7.17 in two different runs. In CUNN mode, the results were identical in two runs (20.8/6.85). Training in CPU mode also gives same results in each run, but not the same as CUNN (19.0/6.99). I got yet different numbers in CPU mode on my laptop (19.6/6.94). Given this, consistency between runs in a particular mode/computer seems like a somewhat arbitrary requirement.

Of course, CUNN is significantly slower than CUDNN although it's still using the GPU. For this example, an epoch took:

1) CUDNN: 3.6 minutes
2) CUNN: 10 minutes
3) CPU: 63/85 minutes (p2.xlarge/my laptop)

Training on full SQuAD takes about 105 minutes per epoch for CUNN (compared to 38 minutes on CUDNN), so about 2.8x slowdown in both cases. FWIW, a CUNN run on the full SQuAD reached EM/F1 67.8/77.4 after epoch 6 (http://beaker.allenai.org/ex/ex_h2ujgatwzgjl, still in progress at epoch 10). I also noticed that even CUNN runs varied between my p2 machine and running in beaker using the docker image, I'm guessing due to different CUDA versions (7.5 vs 8). I did confirm that two separate beaker runs agreed after 2 epochs.
